Add improved Windows UNC path support in std/os

havardjohn opened the pull request:
Original issue: `std/os.createDir` tries to create every component of
the given path as a directory. The problem is that `createDir`
interprets every backslash/slash as a path separator. For a UNC path
this is incorrect. E.g. one UNC form is `\\Server\Volume\Path`. It's an
error to create the `\\Server` directory, as well as creating
`\\Server\Volume`.

Add `ntpath.nim` module with `splitDrive` proc. This implements UNC path
parsing as implemented in the Python `ntpath.py` module. The following
UNC forms are supported:

* `\\Server\Volume\Path`
* `\\?\Volume\Path`
* `\\?\UNC\Server\Volume\Path`

Improves support for UNC paths in various procs in `std/os`:
---

* pathnorm.addNormalizePath
  * Issue: This had incomplete support for UNC paths
    * The UNC prefix (first 2 characters of a UNC path) was assumed to
      be exactly `\\`, but it can be `//` and `\/`, etc. as well
    * Also, the UNC prefix must be normalized to the `dirSep` argument
      of `addNormalizePath`
  * Resolution: Changed to account for different UNC prefixes, and
    normalizing the prefixes according to `dirSep`
    * Affected procs that get tests: `relativePath`, `joinPath`
  * Issue: The server/volume part of UNC paths can be stripped when
    normalizing `..` path components
    * This error should be negligable, so ignoring this
* splitPath
  * Now make sure the UNC drive is not split; return the UNC drive as
    `head` if the UNC drive is the only component of the path
  * Consequently fixes `extractFilename`, `lastPathPart`
* parentDir / `/../`
  * Strip away drive before working on the path, prepending the drive
    after all work is done - prevents stripping UNC components
  * Return empty string if drive component is the only component; this
    is the behavior for POSIX paths as well
  * Alternative implementation: Just call something like
    `pathnorm.normalizePath(path & "/..")` for the whole proc - maybe
    too big of a change
* tailDir
  * If drive is present in path, just split that from path and return
    path
* parentDirs iterator
  * Uses `parentDir` for going backwards
  * When going forwards, first `splitDrive`, yield the drive field, and
    then iterate over path field as normal
* splitFile
  * Make sure path parsing stops at end of drive component
* createDir
  * Fixed by skipping drive part before creating directories
  * Alternative implementation: use `parentDirs` iterator instead of
    iterating over characters
    * Consequence is that it will try to create the root directory
* isRootDir
  * Changed to treat UNC drive alone as root (e.g. "//?/c:" is root)
  * This change prevents the empty string being yielded by the
    `parentDirs` iterator with `fromRoot = false`
* Internal `sameRoot`
  * The "root" refers to the drive, so `splitDrive` can be used here

This adds UNC path support to all procs that could use it in std/os. I
don't think any more work has to be done to support UNC paths. For the
future, I believe the path handling code can be refactored due to
duplicate code. There are multiple ways of manipulating paths, such as
manually searching string for path separator and also having a path
normalizer (pathnorm.nim). If all path manipulation used `pathnorm.nim`,
and path component splitting used `parentDirs` iterator, then a lot of
code could be removed.

Tests
---

Added test file for `pathnorm.nim` and `ntpath.nim`.
`pathnorm.normalizePath` has no tests, so I'm adding a few unit tests.
`ntpath.nim` contains tests copied from Python's test suite.

Added integration tests to `tos.nim` that tests UNC paths.

Removed incorrect `relativePath` runnableExamples from being tested on Windows:
---

`relativePath("/Users///me/bar//z.nim", "//Users/", '/') == "me/bar/z.nim"`

This is incorrect on Windows because the `/` and `//` are not the same
root. `/` (or `\`) is expanded to the drive in the current working
directory (e.g. `C:\`). `//` (or `\\`), however, are the first two
characters of a UNC path. The following holds true for normal Windows
installations:

* `dirExists("/Users") != dirExists("//Users")`
* `dirExists("\\Users") != dirExists("\\\\Users")`

Fixes #19103

Questions:
---

* Should the `splitDrive` proc be in `os.nim` instead with copyright
  notice above the proc?
* Is it fine to put most of the new tests into the `runnableExamples`
  section of the procs in std/os?

Commit:
Add improved Windows UNC path support in std/os

Original issue: `std/os.createDir` tries to create every component of
the given path as a directory. The problem is that `createDir`
interprets every backslash/slash as a path separator. For a UNC path
this is incorrect. E.g. one UNC form is `\\Server\Volume\Path`. It's an
error to create the `\\Server` directory, as well as creating
`\\Server\Volume`.

Add `ntpath.nim` module with `splitDrive` proc. This implements UNC path
parsing as implemented in the Python `ntpath.py` module. The following
UNC forms are supported:

* `\\Server\Volume\Path`
* `\\?\Volume\Path`
* `\\?\UNC\Server\Volume\Path`

Improves support for UNC paths in various procs in `std/os`:
---

* pathnorm.addNormalizePath
  * Issue: This had incomplete support for UNC paths
    * The UNC prefix (first 2 characters of a UNC path) was assumed to
      be exactly `\\`, but it can be `//` and `\/`, etc. as well
    * Also, the UNC prefix must be normalized to the `dirSep` argument
      of `addNormalizePath`
  * Resolution: Changed to account for different UNC prefixes, and
    normalizing the prefixes according to `dirSep`
    * Affected procs that get tests: `relativePath`, `joinPath`
  * Issue: The server/volume part of UNC paths can be stripped when
    normalizing `..` path components
    * This error should be negligable, so ignoring this
* splitPath
  * Now make sure the UNC drive is not split; return the UNC drive as
    `head` if the UNC drive is the only component of the path
  * Consequently fixes `extractFilename`, `lastPathPart`
* parentDir / `/../`
  * Strip away drive before working on the path, prepending the drive
    after all work is done - prevents stripping UNC components
  * Return empty string if drive component is the only component; this
    is the behavior for POSIX paths as well
  * Alternative implementation: Just call something like
    `pathnorm.normalizePath(path & "/..")` for the whole proc - maybe
    too big of a change
* tailDir
  * If drive is present in path, just split that from path and return
    path
* parentDirs iterator
  * Uses `parentDir` for going backwards
  * When going forwards, first `splitDrive`, yield the drive field, and
    then iterate over path field as normal
* splitFile
  * Make sure path parsing stops at end of drive component
* createDir
  * Fixed by skipping drive part before creating directories
  * Alternative implementation: use `parentDirs` iterator instead of
    iterating over characters
    * Consequence is that it will try to create the root directory
* isRootDir
  * Changed to treat UNC drive alone as root (e.g. "//?/c:" is root)
  * This change prevents the empty string being yielded by the
    `parentDirs` iterator with `fromRoot = false`
* Internal `sameRoot`
  * The "root" refers to the drive, so `splitDrive` can be used here

This adds UNC path support to all procs that could use it in std/os. I
don't think any more work has to be done to support UNC paths. For the
future, I believe the path handling code can be refactored due to
duplicate code. There are multiple ways of manipulating paths, such as
manually searching string for path separator and also having a path
normalizer (pathnorm.nim). If all path manipulation used `pathnorm.nim`,
and path component splitting used `parentDirs` iterator, then a lot of
code could be removed.

Tests
---

Added test file for `pathnorm.nim` and `ntpath.nim`.
`pathnorm.normalizePath` has no tests, so I'm adding a few unit tests.
`ntpath.nim` contains tests copied from Python's test suite.

Added integration tests to `tos.nim` that tests UNC paths.

Removed incorrect `relativePath` runnableExamples from being tested on Windows:
---

`relativePath("/Users///me/bar//z.nim", "//Users/", '/') == "me/bar/z.nim"`

This is incorrect on Windows because the `/` and `//` are not the same
root. `/` (or `\`) is expanded to the drive in the current working
directory (e.g. `C:\`). `//` (or `\\`), however, are the first two
characters of a UNC path. The following holds true for normal Windows
installations:

* `dirExists("/Users") != dirExists("//Users")`
* `dirExists("\\Users") != dirExists("\\\\Users")`

Fixes #19103

Questions:
---

* Should the `splitDrive` proc be in `os.nim` instead with copyright
  notice above the proc?
* Is it fine to put most of the new tests into the `runnableExamples`
  section of the procs in std/os?

Varriount commented on lib/pure/os.nim:
```diff
@@ -257,9 +260,16 @@ proc splitPath*(path: string): tuple[head, tail: string] {.
+    let (drive, splitpath) = splitDrive(path)
+    let stop = drive.len
+  else:
+    const stop = 0
```
```suggestion
    const stop = 0

```

Varriount commented on lib/pure/os.nim:
```diff
@@ -330,40 +341,8 @@ when doslikeFileSystem:
-          return h == 2
-        else:
-          return false
+    result = (isAbsFromCurrentDrive(path1) and isAbsFromCurrentDrive(path2)) or
+      cmpIgnoreCase(splitDrive(path1).drive, splitDrive(path2).drive) == 0
```
I know this is longer, but I feel that this is a bit easier to read - it's not as "information-dense". What do you think?
```suggestion
    if isAbsFromCurrentDrive(path1) and isAbsFromCurrentDrive(path2):
      result = true
    elif cmpIgnoreCase(splitDrive(path1).drive, splitDrive(path2).drive) == 0:
      result = true
    else:
      result = false
```

havardjohn replied:
I think you have a good point.

Varriount commented on lib/pure/os.nim:
```diff
@@ -384,11 +363,14 @@ proc relativePath*(path, base: string, sep = DirSep): string {.
     assert relativePath("/Users/me/bar/z.nim", "/Users/other/bad", '/') == "../../me/bar/z.nim"
     assert relativePath("/Users/me/bar/z.nim", "/Users/other", '/') == "../me/bar/z.nim"
-    assert relativePath("/Users///me/bar//z.nim", "//Users/", '/') == "me/bar/z.nim"
+    when not doslikeFileSystem: # `//` means a UNC-path on Windows
```
```suggestion
    when not doslikeFileSystem: # On Windows, UNC-paths start with `//`
```

Varriount commented on lib/pure/os.nim:
```diff
@@ -496,7 +478,17 @@ proc parentDir*(path: string): string {.
+      assert parentDir("\\\\?\\c:") == ""
+      assert parentDir("//?/c:/Users") == "\\\\?\\c:"
+      assert parentDir("\\\\localhost\\c$") == ""
+      assert parentDir("\\Users") == "\\"
```
```suggestion
      assert parentDir(r"\\?\c:") == r""
      assert parentDir(r"//?/c:/Users") == r"\\?\c:"
      assert parentDir(r"\\\\localhost\\c$") == r""
      assert parentDir(r"\Users") == r"\"
```

Varriount commented on lib/pure/os.nim:
```diff
@@ -524,8 +523,16 @@ proc tailDir*(path: string): string {.
+    when doslikeFileSystem:
+      assert tailDir("//?/c:") == ""
+      assert tailDir("//?/c:/Users") == "Users"
+      assert tailDir("\\\\localhost\\c$\\Windows\\System32") == "Windows\\System32"
```
```suggestion
      assert tailDir(r"\\localhost\c$\Windows\System32") == r"Windows\System32"
```

Varriount commented on lib/pure/os.nim:
```diff
@@ -543,7 +550,14 @@ proc isRootDir*(path: string): bool {.
+    when doslikeFileSystem:
+      assert isRootDir("//?/c:")
+      assert isRootDir("//?/UNC/localhost/c$")
+      assert not isRootDir("\\\\?\\c:\\Users")
```
```suggestion
      assert not isRootDir(r"\\?\c:\Users")
```

Varriount commented on lib/pure/os.nim:
```diff
@@ -580,6 +594,22 @@ iterator parentDirs*(path: string, fromRoot=false, inclusive=true): string =
+      doAssert parentDirs("//?/UNC/localhost/c$/Users", fromRoot = false).toSeq ==
+        @["//?/UNC/localhost/c$/Users", "\\\\?\\UNC\\localhost\\c$"]
+      doAssert parentDirs("\\Users", fromRoot = true).toSeq == @["\\", "\\Users"]
+      doAssert parentDirs("\\Users", fromRoot = false).toSeq == @["\\Users", "\\"]
```
```suggestion
      doAssert parentDirs(r"C:\Users", fromRoot = true).toSeq == @[r"C:\", r"C:\Users"]
      doAssert parentDirs(r"C:\Users", fromRoot = false).toSeq == @[r"C:\Users", r"C:"]
      doAssert parentDirs(r"\\?\c:\Users", fromRoot = true).toSeq ==
        @[r"\\?\c:\", r"\\?\c:\Users"]
      doAssert parentDirs(r"\\?\c:\Users", fromRoot = false).toSeq ==
        @[r"\\?\c:\Users", r"\\?\c:"]
      doAssert parentDirs(r"//localhost/c$/Users", fromRoot = true).toSeq ==
        @[r"//localhost/c$/", r"//localhost/c$/Users"]
      doAssert parentDirs(r"//?/UNC/localhost/c$/Users", fromRoot = false).toSeq ==
        @[r"//?/UNC/localhost/c$/Users", r"\\?\UNC\localhost\c$"]
      doAssert parentDirs(r"\Users", fromRoot = true).toSeq == @[r"\", r"\Users"]
      doAssert parentDirs(r"\Users", fromRoot = false).toSeq == @[r"\Users", r"\"]
```

Varriount commented on lib/pure/os.nim:
```diff
@@ -607,12 +641,20 @@ proc `/../`*(head, tail: string): string {.noSideEffect.} =
+    when doslikeFileSystem:
+      assert "//?/c:" /../ "d/e" == "\\\\?\\c:\\d\\e"
+      assert "//?/c:/Users" /../ "d/e" == "\\\\?\\c:\\d\\e"
+      assert "\\\\localhost\\c$" /../ "d/e" == "\\\\localhost\\c$\\d\\e"
```
```suggestion
      assert r"//?/c:" /../ "d/e" == r"\\?\c:\d\e"
      assert r"//?/c:/Users" /../ "d/e" == r"\\?\c:\d\e"
      assert r"\\localhost\c$" /../ "d/e" == r"\\localhost\c$\d\e"
```

Varriount commented on lib/pure/os.nim:
```diff
@@ -677,10 +719,23 @@ proc splitFile*(path: string): tuple[dir, name, ext: string] {.
+      assert splitFile("//?/c:") == ("//?/c:", "", "")
+      assert splitFile("//?/c:/Users") == ("//?/c:", "Users", "")
+      (dir, name, ext) = splitFile("\\\\localhost\\c$\\test.txt")
+      assert dir == "\\\\localhost\\c$"
```
```suggestion
      (dir, name, ext) = splitFile(r"\\localhost\c$\test.txt")
      assert dir == r"\\localhost\c$"
```

Varriount commented on lib/std/private/ntpath.nim:
```diff
@@ -0,0 +1,64 @@
+    doAssert splitDrive("C:") == ("C:", "")
+    doAssert splitDrive("C:\\") == ("C:", "\\")
+    doAssert splitDrive("\\\\server\\drive\\foo\\bar") == ("\\\\server\\drive", "\\foo\\bar")
+    doAssert splitDrive("\\\\?\\UNC\\server\\share\\dir") == ("\\\\?\\UNC\\server\\share", "\\dir")
```
```suggestion
    doAssert splitDrive(r"C:\") == (r"C:", r"\")
    doAssert splitDrive(r"\\server\drive\foo\bar") == (r"\\server\drive", r"\foo\bar")
    doAssert splitDrive(r"\\?\UNC\server\share\dir") == (r"\\?\UNC\server\share", r"\dir")
```

Varriount commented on tests/stdlib/tntpath.nim:
```diff
@@ -0,0 +1,53 @@
+  # Special for `\`-rooted paths on Windows. I don't know if this is correct,
+  # but `\` is not recognized as a drive, in contrast to `C:` or `\\?\c:`.
+  # This behavior is the same for Python's `splitdrive` function.
+  doAssert splitDrive("\\\\") == ("", "\\\\")
```
```suggestion
  doAssert splitDrive(r"c:\foo\bar") == (r"c:", r"\foo\bar")
  doAssert splitDrive(r"c:/foo/bar") == (r"c:", r"/foo/bar")
  doAssert splitDrive(r"\\conky\mountpoint\foo\bar") == (r"\\conky\mountpoint", r"\foo\bar")
  doAssert splitDrive(r"//conky/mountpoint/foo/bar") == (r"//conky/mountpoint", r"/foo/bar")
  doAssert splitDrive(r"\\\conky\mountpoint\foo\bar") == (r"", r"\\\conky\mountpoint\foo\bar")
  doAssert splitDrive(r"///conky/mountpoint/foo/bar") == (r"", r"///conky/mountpoint/foo/bar")
  doAssert splitDrive(r"\\conky\\mountpoint\foo\bar") == (r"", r"\\conky\\mountpoint\foo\bar")
  doAssert splitDrive(r"//conky//mountpoint/foo/bar") == (r"", r"//conky//mountpoint/foo/bar")
  # Issue #19911: UNC part containing U+0130
  doAssert splitDrive(r"//conky/MOUNTPOİNT/foo/bar") == (r"//conky/MOUNTPOİNT", r"/foo/bar")
  # gh-81790: support device namespace, including UNC drives.
  doAssert splitDrive(r"//?/c:") == (r"//?/c:", r"")
  doAssert splitDrive(r"//?/c:/") == (r"//?/c:", r"/")
  doAssert splitDrive(r"//?/c:/dir") == (r"//?/c:", r"/dir")
  doAssert splitDrive(r"//?/UNC") == (r"", r"//?/UNC")
  doAssert splitDrive(r"//?/UNC/") == (r"", r"//?/UNC/")
  doAssert splitDrive(r"//?/UNC/server/") == (r"//?/UNC/server/", r"")
  doAssert splitDrive(r"//?/UNC/server/share") == (r"//?/UNC/server/share", r"")
  doAssert splitDrive(r"//?/UNC/server/share/dir") == (r"//?/UNC/server/share", r"/dir")
  doAssert splitDrive(r"//?/VOLUME{00000000-0000-0000-0000-000000000000}/spam") == (r"//?/VOLUME{00000000-0000-0000-0000-000000000000}", r"/spam")
  doAssert splitDrive(r"//?/BootPartition/") == (r"//?/BootPartition", r"/")

  doAssert splitDrive(r"\\?\c:") == (r"\\?\c:", r"")
  doAssert splitDrive(r"\\?\c:\") == (r"\\?\c:", r"\")
  doAssert splitDrive(r"\\?\c:\dir") == (r"\\?\c:", r"\dir")
  doAssert splitDrive(r"\\?\UNC") == (r"", r"\\?\UNC")
  doAssert splitDrive(r"\\?\UNC\") == (r"", r"\\?\UNC\")
  doAssert splitDrive(r"\\?\UNC\server\") == (r"\\?\UNC\server\", r"")
  doAssert splitDrive(r"\\?\UNC\server\share") == (r"\\?\UNC\server\share", r"")
  doAssert splitDrive(r"\\?\UNC\server\share\dir") == (r"\\?\UNC\server\share", r"\dir")
  doAssert splitDrive(r"\\?\VOLUME{00000000-0000-0000-0000-000000000000}\spam") == (r"\\?\VOLUME{00000000-0000-0000-0000-000000000000}", r"\spam")
  doAssert splitDrive(r"\\?\BootPartition\") == (r"\\?\BootPartition", r"\")

block:
  doAssert splitDrive(r"C:") == (r"C:", r"")
  doAssert splitDrive(r"C:\") == (r"C:", r"\")
  doAssert splitDrive(r"non/absolute/path") == (r"", r"non/absolute/path")

  # Special for `\`-rooted paths on Windows. I don't know if this is correct,
  # rbut `\` is not recognized as a drive, in contrast to `C:` or `\?\c:`.
  # This behavior is the same for Python's `splitdrive` function.
  doAssert splitDrive(r"\\") == (r"", r"\\")
```

Varriount commented on tests/stdlib/tpathnorm.nim:
```diff
@@ -0,0 +1,34 @@
+
+  block: # / -> \
+    initVars
+    addNormalizePath("//?/c:/./foo//bar/../baz", result, state, '\\')
+    doAssert result == "\\\\?\\c:\\foo\\baz"
+    addNormalizePath("me", result, state, '\\')
+    doAssert result == "\\\\?\\c:\\foo\\baz\\me"
```
```suggestion
    addNormalizePath(r"//?/c:/./foo//bar/../baz", result, state, '\\')
    doAssert result == r"\\?\c:\foo\baz"
    addNormalizePath("me", result, state, '\\')
    doAssert result == r"\\?\c:\foo\baz\me"
```

Varriount commented on tests/stdlib/tpathnorm.nim:
```diff
@@ -0,0 +1,34 @@
+
+  block: # Append path component to UNC drive
+    initVars
+    addNormalizePath("//?/c:", result, state, '\\')
+    doAssert result == "\\\\?\\c:"
+    addNormalizePath("Users", result, state, '\\')
+    doAssert result == "\\\\?\\c:\\Users"
+    addNormalizePath("me", result, state, '\\')
+    doAssert result == "\\\\?\\c:\\Users\\me"
```
```suggestion
    addNormalizePath(r"//?/c:", result, state, '\\')
    doAssert result == r"\\?\c:"
    addNormalizePath("Users", result, state, '\\')
    doAssert result == r"\\?\c:\Users"
    addNormalizePath("me", result, state, '\\')
    doAssert result == r"\\?\c:\Users\me"
```

Araq commented:
Thank you! Additional tests should go into `tests/stdlib/tos.nim` though.

Commit:
[skipci] Apply suggestions from code review

Co-authored-by: Clay Sweetser <Varriount@users.noreply.github.com>

Commit:
[skip ci] Update lib/pure/os.nim

Co-authored-by: Clay Sweetser <Varriount@users.noreply.github.com>

havardjohn commented:
@Araq Alright, I'll move the runnableExamples tests into tos.nim.

Commit:
Move runnableExamples tests in os.nim to tos.nim

havardjohn commented:
Moved and fixed tests. @Varriount Thanks for the review

Commit:
Merge remote-tracking branch 'origin/devel' into feature/improve_unc_path_support

Commit:
tests/topt_no_cursor: Change from using splitFile to splitDrive

`splitFile` can no longer be used in the test, because it generates
different ARC code on Windows and Linux. This replaces `splitFile` with
`splitDrive`, because it generates same ARC code on Windows and Linux,
and returns a tuple. I assume the test wants a proc that returns a
tuple.

havardjohn commented:
Looking at CI

tests/misc/t19600.nim seems to be flaky. Locally, it can both succeed and fail.
tests/async/tasyncssl.nim seems to be flaky on mac.

tests/arc/topt_no_cursor.nim failing on Windows is a problem. It uses `splitFile`, which now has a different body on Windows than Linux, and generates different code. I think a different `splitDrive` can be used instead. I have pushed a commit with this change. Is this change fine?

Varriount commented:
Can you explain in more detail why the tests are failing, and what the planned changes are?

havardjohn commented:
@Varriount
tests/misc/t19600.nim: I don't know why this fails. It fails on devel as well with the same frequency.
tests/async/tasyncssl.nim: I don't know why this fails. I don't have a macos environment. Want to see if it succeeds now. This test doesn't pull in any of this PR's changes.
tests/arc/topt_no_cursor.nim: I detailed reasoning for my changes in commit message.

Edit: I will try to setup a macos environment and find out what to do about tasyncssl.nim

Varriount commented:
Note - some of these tests may just be flaky in general, and may have nothing to do with the changes in this PR.

Araq commented on lib/std/private/ntpath.nim:
```diff
@@ -0,0 +1,64 @@
+# This module contains derivative work of Python's `ntpath.py` module. This is
+# attributed to the PSF-2.0 license with the given notice:
+#
+# Copyright (c) 2001-2022 Python Software Foundation; All Rights Reserved
```
This makes me nervous... :-)

The PSF does not hold the rights on this Nim code that was only inspired by some Python code. As far as I can tell.

havardjohn replied:
After reading this: https://softwareengineering.stackexchange.com/a/234829, I decided to add this copyright notice. The `splitdrive` function in Python is PSF-2.0-licensed as far as I understood the Python license. PSF-2.0 is a BSD-style license.

Also, `lib/std/private/win_setenv.nim` in Nim has a copyright notice to Facebook at the top. I don't have an interest in licensing, so I would like to be wrong. Should I just remove the copyright stuff in ntpath.nim and tntpath.nim?

Varriount replied:
While do I think that some form of attribution should be given, I don't feel that code would fall under the PSF's copyright. This is mainly because the `splitDrive` algorithm is fairly "obvious" - there's only so many ways one could implement such logic. Additionally, I've seen this kind of thing before, where an algorithms in one project contains attribution to the same algorithm in another project; sometimes the two projects are written in the same language, sometimes they are written in another language.

Of course, if the PSF comes along and says, "that code is a derivative work, and should be under the PSF license", then we can discuss that with them. Until then, a simple "inspired by <url to file>" is good enough.

havardjohn replied:
Thanks for the explanation. I dropped the attribution to Python.

Commit:
Merge remote-tracking branch 'origin/devel' into feature/improve_unc_path_support

Commit:
Drop copyright attribute to Python
